dsl/src/main/scala/org/http4s/dsl/Path.scala

```diff
@@ -249,6 +249,48 @@ abstract class OptionalQueryParamDecoderMatcher[T: QueryParamDecoder](name: Stri
     ).toOption
 }
 
+/**
+  * Capture a query parameter that appears 0 or more times.
+  *
+  * {{{
+  *   case class Foo(i: Int)
+  *   implicit val fooDecoder: QueryParamDecoder[Foo] = ...
+  *   implicit val fooParam: QueryParam[Foo] = ...
+  *
+  *   object FooMatcher extends OptionalMultiQueryParamDecoderMatcher[Foo]("foo")
+  *   val service: HttpService = {
+  *     // matches http://.../closest?foo=2&foo=3&foo=4
+  *     case GET -> Root / "closest" :? FooMatcher(Some(Seq(2,3,4))) => ...
+  *
+  *     /*
+  *     *  matches http://.../closest?foo=2&foo=3&foo=4 as well as http://.../closest (no parameters)
+  *     *  or http://.../closest?foo=2 (single occurrence)
+  *     */
+  *     case GET -> Root / "closest" :? FooMatcher(is) => ...
+  * }}}
+  */
+abstract class OptionalMultiQueryParamDecoderMatcher[T: QueryParamDecoder](name: String) {
+  import scalaz.{ValidationNel, Success, Failure}
+
+  def unapply(params: Map[String, Seq[String]]): Option[Option[ValidationNel[ParseFailure, Seq[T]]]] = {
```

Comment:
Why an `Option[Option[ValidationNel[ParseFailure, Seq[T]]]]` instead of an `Option[ValidationNel[ParseFailure, Seq[T]]`?  I think `Success(Nil)` would be legit for an optional many-matcher.

I've come to think of `Seq` as a smell, since it can be mutable.  How about a `List` or a `Vector` instead of a `Seq`?  (I realize `params` are a `Seq`. I'd like to fix that eventually as well.)